### ISSUE TITLE
Can't get attribute 'Silence' on <module 'ultralytics.nn.modules.block' >

### ISSUE THREAD

**Issue by ssuncheol** [ultralytics/ultralytics#8667]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to proceed with object detection using yolov9, and there are the following issues. Is it an ultralytics version issue?

**Comment by glenn-jocher** [comment]
@ssuncheol hello! It seems like you're encountering an issue related to a missing attribute `Silence` in the `ultralytics.nn.modules.block` module. This could potentially be due to a version mismatch or an incomplete installation.

First, ensure you're using the correct version of YOLOv8, as you mentioned YOLOv9, which might have been a typo since the current focus is on YOLOv8. If you're working with the latest features or models, it's always a good idea to update to the latest version of the Ultralytics repository. You can do so by running:

```bash
pip install --upgrade ultralytics
```

If the problem persists, it might help to completely reinstall the package to ensure all components are correctly installed:

```bash
pip uninstall ultralytics
pip install ultralytics
```

If you continue to face issues, please provide more details about your environment (e.g., Python version, Ultralytics version) and the exact steps or code leading to this error. This will help us diagnose the problem more effectively. 🛠️

**Comment by sainisatish** [comment]
I am getting below error.

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py:714, in torch_safe_load(weight)
    707     with temporary_modules(
    708         {
    709             "ultralytics.yolo.utils": "ultralytics.utils",
   (...)
    712         }
    713     ):  # for legacy 8.0 Classify and Pose models
--> 714         ckpt = torch.load(file, map_location="cpu")
    716 except ModuleNotFoundError as e:  # e.name is missing module name

File /opt/conda/lib/python3.10/site-packages/torch/serialization.py:1014, in load(f, map_location, pickle_module, weights_only, mmap, **pickle_load_args)
   1013                 raise pickle.UnpicklingError(UNSAFE_MESSAGE + str(e)) from None
-> 1014         return _load(opened_zipfile,
   1015                      map_location,
   1016                      pickle_module,
   1017                      overall_storage=overall_storage,
   1018                      **pickle_load_args)
   1019 if mmap:

File /opt/conda/lib/python3.10/site-packages/torch/serialization.py:1422, in _load(zip_file, map_location, pickle_module, pickle_file, overall_storage, **pickle_load_args)
   1421 unpickler.persistent_load = persistent_load
-> 1422 result = unpickler.load()
   1424 torch._utils._validate_loaded_sparse_tensors()

File /opt/conda/lib/python3.10/site-packages/torch/serialization.py:1415, in _load.<locals>.UnpicklerWrapper.find_class(self, mod_name, name)
   1414 mod_name = load_module_mapping.get(mod_name, mod_name)
-> 1415 return super().find_class(mod_name, name)

ModuleNotFoundError: No module named 'models'

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
Cell In[23], line 1
----> 1 model = YOLO("/kaggle/working/TrainedModel/best.pt")

File /opt/conda/lib/python3.10/site-packages/ultralytics/models/yolo/model.py:23, in YOLO.__init__(self, model, task, verbose)
     20     self.__dict__ = new_instance.__dict__
     21 else:
     22     # Continue with default YOLO initialization
---> 23     super().__init__(model=model, task=task, verbose=verbose)

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/model.py:142, in Model.__init__(self, model, task, verbose)
    140     self._new(model, task=task, verbose=verbose)
    141 else:
--> 142     self._load(model, task=task)
    144 self.model_name = model

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/model.py:229, in Model._load(self, weights, task)
    227 suffix = Path(weights).suffix
    228 if suffix == ".pt":
--> 229     self.model, self.ckpt = attempt_load_one_weight(weights)
    230     self.task = self.model.args["task"]
    231     self.overrides = self.model.args = self._reset_ckpt_args(self.model.args)

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py:788, in attempt_load_one_weight(weight, device, inplace, fuse)
    786 def attempt_load_one_weight(weight, device=None, inplace=True, fuse=False):
    787     """Loads a single model weights."""
--> 788     ckpt, weight = torch_safe_load(weight)  # load ckpt
    789     args = {**DEFAULT_CFG_DICT, **(ckpt.get("train_args", {}))}  # combine model and default args, preferring model args
    790     model = (ckpt.get("ema") or ckpt["model"]).to(device).float()  # FP32 model

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py:718, in torch_safe_load(weight)
    716 except ModuleNotFoundError as e:  # e.name is missing module name
    717     if e.name == "models":
--> 718         raise TypeError(
    719             emojis(
    720                 f"ERROR ❌️ {weight} appears to be an Ultralytics YOLOv5 model originally trained "
    721                 f"with [https://github.com/ultralytics/yolov5.](https://github.com/ultralytics/yolov5.%3C/span%3E%3Cspan) class="ansi-bold" style="color:rgb(175,95,0)">\nThis model is NOT forwards compatible with "
    722                 f"YOLOv8 at https://github.com/ultralytics/ultralytics. style="color:rgb(175,0,0)">"
    723                 f"\nRecommend fixes are to train a new model using the latest 'ultralytics' package or to "
    724                 f"run a command with an official YOLOv8 model, i.e. 'yolo predict model=yolov8n.pt'"
    725             )
    726         ) from e
    727     LOGGER.warning(
    728         f"WARNING ⚠️ {weight} appears to require '{e.name}', which is not in ultralytics requirements."
    729         f"\nAutoInstall will run now for '{e.name}' but this feature will be removed in the future."
    730         f"\nRecommend fixes are to train a new model using the latest 'ultralytics' package or to "
    731         f"run a command with an official YOLOv8 model, i.e. 'yolo predict model=yolov8n.pt'"
    732     )
    733     check_requirements(e.name)  # install missing module

TypeError: ERROR ❌️ /kaggle/working/TrainedModel/best.pt appears to be an Ultralytics YOLOv5 model originally trained with https://github.com/ultralytics/yolov5.
This model is NOT forwards compatible with YOLOv8 at https://github.com/ultralytics/ultralytics.
Recommend fixes are to train a new model using the latest 'ultralytics' package or to run a command with an official YOLOv8 model, i.e. 'yolo predict model=yolov8n.pt'

**Comment by glenn-jocher** [comment]
@sainisatish hey there! It looks like you're trying to load a YOLOv5 model with the Ultralytics YOLOv8 package, which unfortunately isn't compatible due to differences in the model architectures. 😅

To resolve this, you have a couple of options:
1. Train a new model using the latest `ultralytics` package for YOLOv8 compatibility.
2. Use an official YOLOv8 model for your tasks, for example:
```bash
yolo predict model=yolov8n.pt source='your_image_or_video_path'
```

This should get you up and running with YOLOv8! If you're specifically looking to work with YOLOv5 models, consider using the YOLOv5 repository and package. Happy detecting! 🚀

**Comment by sainisatish** [comment]
> @sainisatish hey there! It looks like you're trying to load a YOLOv5 model with the Ultralytics YOLOv8 package, which unfortunately isn't compatible due to differences in the model architectures. 😅
> 
> To resolve this, you have a couple of options:
> 
> 1. Train a new model using the latest `ultralytics` package for YOLOv8 compatibility.
> 2. Use an official YOLOv8 model for your tasks, for example:
> 
> ```shell
> yolo predict model=yolov8n.pt source='your_image_or_video_path'
> ```
> 
> This should get you up and running with YOLOv8! If you're specifically looking to work with YOLOv5 models, consider using the YOLOv5 repository and package. Happy detecting! 🚀

I Have trained the Yolov9 model but when I am trying to  loading this model  for inference using below code
`from ultralytics import YOLO
model = YOLO("yolov9.pt")`

it is throwing the above mention error.

**Comment by Friman04** [comment]
same

**Comment by theShamrockCoder** [comment]
Same

**Comment by glenn-jocher** [comment]
This error is because the model was not trained with the ultralytics Python package. As mentioned in my post above:

> * Train a new model using the latest `ultralytics` package for best compatibility (RECOMMENDED).

**Comment by geoxpert0001** [comment]
So
Did we train v9.pt using Ultralytics and make predictions with it? 
Did we not support using YOLOv9 for training and obtaining weights for prediction with Ultralytics?"

**Comment by glenn-jocher** [comment]
@geoxpert0001 hey there! 😊 Yes, for using custom weights trained with WongKinYiu/yolov9 for prediction in the Ultralytics framework, it's recommended to train your model using the Ultralytics method for compatibility. Unfortunately, directly loading weights trained outside of the Ultralytics ecosystem into the Ultralytics YOLOv8 package might not work due to differences in model architecture and training methods.

```python
# This might not work as expected
model = YOLO("use_WongKinYiu/yolov9_train.pt")
```

For the best experience and compatibility, retraining with the Ultralytics package or using official Ultralytics models is the way to go. Happy coding! 🚀

**Comment by ebayes** [comment]
Hey - I trained with the Ultralytics package (see below) but am still getting the 'Silence' error. Any ideas why?


This was my training script:

yolo = YOLO('yolov9e.pt')
yolo.train(data='/path/to/data.yaml', epochs=300)


This is my inference script:

model = YOLO('/path/to/custom_yolov9e_weights.pt')
results = model('path/to/image.jpg')

**Comment by glenn-jocher** [comment]
@ebayes hey there! 😊 It looks like there's a little mix-up with the model naming in your scripts. When you train a model, it outputs `.pt` files named differently (e.g., `best.pt`, `last.pt`) in your training folder. Ensure you're using the correct path to these weight files during inference. Your inference script should look something like:

```python
model = YOLO('/path/to/training/folder/best.pt')  # Use the exact name like `best.pt` or `last.pt`
results = model.predict(source='path/to/image.jpg')  # Use `predict` method for inference
```
Double-check the path to your custom weights and use the `predict` method for running inference. Hope this helps! 🚀

**Comment by ebayes** [comment]
Thanks Glenn! I actually changes the name of the weights so I don't think it's that. It's to do with the model itself. I trained 9e. Perhaps it's an issue with 9?

**Comment by glenn-jocher** [comment]
@ebayes hello! It sounds like there might be a compatibility issue or a specific problem related to the `yolov9e` model implementation. Could you please check if the model version you trained is fully supported by the version of the Ultralytics framework you are using? Sometimes, specific models or features require the very latest updates. Running a quick update often helps:

```bash
pip install --upgrade ultralytics
```

If you're still encountering issues after ensuring everything is up-to-date, it might be helpful to revisit the exact training command and the environment setup to rule out any discrepancies. If the problem persists, providing logs or error messages would be immensely helpful for further troubleshooting! 😊

**Comment by silvada95** [comment]
> @ebayes hello! It sounds like there might be a compatibility issue or a specific problem related to the `yolov9e` model implementation. Could you please check if the model version you trained is fully supported by the version of the Ultralytics framework you are using? Sometimes, specific models or features require the very latest updates. Running a quick update often helps:
> 
> ```shell
> pip install --upgrade ultralytics
> ```
> 
> If you're still encountering issues after ensuring everything is up-to-date, it might be helpful to revisit the exact training command and the environment setup to rule out any discrepancies. If the problem persists, providing logs or error messages would be immensely helpful for further troubleshooting! 😊

I think it is a problem with the Save or Load functions from the Ultralytics repo. For example, when I run "get_model" with the yolov9 yaml file in a separate file, save it with "torch.save", and load it in another script with "torch.load", it works fine. However, when saved with the framework functions, it presents the aforementioned problems...

**Comment by glenn-jocher** [comment]
Hello @leah-at-lovelace,

Thank you for bringing this to our attention! It sounds like you've encountered a regression in the latest update. To help us investigate and resolve this issue, could you please provide a minimum reproducible code example? This will allow us to reproduce the bug on our end. You can find guidelines for creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you are using the latest versions of `torch` and `ultralytics`. If not, kindly upgrade your packages and try again:

```bash
pip install --upgrade torch ultralytics
```

Looking forward to your response so we can get this sorted out for you! 😊

**Comment by leah-at-lovelace** [comment]
I poked around a little more, and it seems like the version of `yolov9e-seg.pt` I downloaded last week is not compatible with ultralytics 8.2.35, but if I download a new version of the same file, using `model = YOLO('yolov9e-seg.pt')`, then it does work. This was pretty unexpected, I assumed the pre-trained weights file would not change between versions.

The difference seems to have been introduced in 8.2.35, if I go back to 8.2.34 the old version works.

Another interesting finding: versions before 8.2.35 are able to read both the current yolov9e-seg.pt weights and the old yolov9e-seg.pt weights, but 8.2.35 can only read the current yolov9e-seg.pt weights.

Here is a minimum reproducible example.
```
from ultralytics import YOLO
model = YOLO('/path/to/yolov9e-seg.pt')
model.train(data='data.yaml', epochs=1)
```

Here is the full error message:
```
Overriding model.yaml nc=80 with nc=1

                   from  n    params  module                                       arguments                     
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/<path>/.conda/envs/lovelace-env/lib/python3.10/site-packages/ultralytics/engine/model.py", line 657, in train
    self.trainer.model = self.trainer.get_model(weights=self.model if self.ckpt else None, cfg=self.model.yaml)
  File "/<path>/.conda/envs/lovelace-env/lib/python3.10/site-packages/ultralytics/models/yolo/segment/train.py", line 34, in get_model
    model = SegmentationModel(cfg, ch=3, nc=self.data["nc"], verbose=verbose and RANK == -1)
  File "/<path>/.conda/envs/lovelace-env/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 372, in __init__
    super().__init__(cfg=cfg, ch=ch, nc=nc, verbose=verbose)
  File "/<path>/.conda/envs/lovelace-env/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 288, in __init__
    self.model, self.save = parse_model(deepcopy(self.yaml), ch=ch, verbose=verbose)  # model, savelist
  File "/<path>/.conda/envs/lovelace-env/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 860, in parse_model
    m = getattr(torch.nn, m[3:]) if "nn." in m else globals()[m]  # get module
KeyError: 'Silence'
```

**Comment by glenn-jocher** [comment]
Hello @leah-at-lovelace-lovelace,

Thank you for providing a detailed report and a minimum reproducible example! 😊

It seems like there might be a compatibility issue introduced in version 8.2.35 of the Ultralytics package. To help us investigate this further, could you please confirm the following:

1. **Package Versions**: Ensure you are using the latest versions of `torch` and `ultralytics`. You can update them using:
   ```bash
   pip install --upgrade torch ultralytics
   ```

2. **Reproducibility**: You've mentioned that the issue occurs with the `yolov9e-seg.pt` weights file downloaded last week. Could you try downloading the latest version of the weights file again and see if the issue persists?

Here's a quick check you can perform:
```python
from ultralytics import YOLO

# Load the model with the latest weights
model = YOLO('path/to/latest/yolov9e-seg.pt')

# Run a simple training command
model.train(data='data.yaml', epochs=1)
```

If the issue persists even after updating the packages and using the latest weights, it might indicate a regression in the recent update. In that case, we will escalate this to our development team for a closer look.

Thank you for your patience and cooperation. We're here to help you get this resolved as quickly as possible! 🚀

**Comment by leah-at-lovelace** [comment]
Thanks for looking into this. I did confirm that I'm using the latest version of torch and ultralytics. With these latest versions it does work with a fresh download of the weights, but does not work with the weights I downloaded last week. So at the end of the day I'm not blocked, and I guess it's not exactly broken, but it was unexpected and I suspect unintended. If nothing else, I hope this serves to document the problem for anyone else who runs into it.
